### PR TITLE
Add read/write to cache + Android instrumentation test

### LINKED ISSUE #35: Cache map data
Along with #21 which downloads map data and inserts it into the database to be indexed for searching, we also want to cache the full set of data so that it can be loaded directly onto the map. To do this, on Android and iOS it should be saved as a file on-device, and in JS it can just be kept in memory (it only needs to exist as long as the web page is open).

# Technical Details

Create a new class using the expect/actual mechanism that contains a function that will handle reading and writing JSON data.
* For Android and iOS: write data to a file, and also read data from that file
* For JS: save (write) the JSON data in a variable, and return (read) the value of that variable
There are existing Kotlin Multiplatform libraries for doing file I/O on Android and iOS that can be used for this part of the task, but it's entirely optional. The `actual` implementations can simply call the native methods to read and write files.

This class should also track the last time the cached data was updated. This will be used to determine whether or not the cache should be updated by comparing the cache's time with the `Last-Modified` header of the remote API response. This will require adding a dependency to a date/time library, such as [this one](https://github.com/Kotlin/kotlinx-datetime).

Finally, update the data repository class to write data to this cache when the data is updated.

### PR DESCRIPTION
**Pull request by safa-howaid**
Closes #35

### CODE REVIEW

File: common/src/commonMain/kotlin/com/cuhacking/atlas/common/FeatureApi.kt [hunk 1]
```diff
@@ -21,10 +23,13 @@ class FeatureApi(private val database: Database, private val client: HttpClient)
         this)
 
     suspend fun getAndStoreFeatures() = withContext(CoroutineDispatchers.io) {
-        val features: List<Feature> = client.get<String>(AtlasConfig.SERVER_URL).toFeatureCollection().features
+        val response = client.get<String>(AtlasConfig.SERVER_URL)
```
Comment (dellisd) [inline common/src/commonMain/kotlin/com/cuhacking/atlas/common/FeatureApi.kt:26]: Be careful to handle possible network errors that could happen here. It probably needs to be wrapped in a try/catch.

File: common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import platform.Foundation.*
+
+actual class DataCache {
+    actual var lastModified: Instant? = null
+    private val cacheDirectory = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, true).first()
+    private val file = (cacheDirectory as NSString).stringByAppendingPathComponent("mapdata.json")
+
+    @Suppress("CAST_NEVER_SUCCEEDS")
```
Comment (dellisd) [inline common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:13]: What is this for? (just checking)

File: common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+import android.content.Context
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import java.io.File
+
+actual class DataCache {
+    lateinit var appContext: Context
+    actual var lastModified: Instant? = null
```
Comment (dellisd) [inline common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:11]:
```suggestion
    actual var lastModified: Instant? = null
        private set
```
We don't want the `lastModified` time to be modified from outside.

**Review by dellisd — COMMENTED** [review]
Looking good!

File: common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt (hunk 2 above)
Comment (safa-howaid) [reply common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:13]: This is for the warning that happens when the Kotlin String is casted to an NSString on line 16. Not sure why it gives a warning. As far as I know, casting a string to an NSString using this syntax is allowed.

Comment (dellisd) [reply common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:13]: You're right, it should be fine. Ah well. Probably a bug.

File: common/src/commonMain/kotlin/com/cuhacking/atlas/common/DataCache.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.cuhacking.atlas.common
+
+import kotlinx.datetime.Instant
+
+expect class DataCache() {
```
[the remark this answers is not in the document]
Comment (safa-howaid) [reply common/src/commonMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:5]: I get an error here that says "Expected class constructor cannot have a property parameter"

[the remark this answers is not in the document]
Comment (safa-howaid) [reply common/src/commonMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:5]: nvm I got rid of val and it works fine!

File: common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import android.content.Context
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import java.io.File
+
+actual class DataCache actual constructor(dispatchers: CoroutineDispatchers) {
+    lateinit var appContext: Context
+    actual var lastModified: Instant? = null
+    private val dispatchers = CoroutineDispatchers
```
Comment (dellisd) [inline common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]: `dispatchers` can be made a constructor property.
```suggestion
actual class DataCache actual constructor(private val dispatchers: CoroutineDispatchers = CoroutineDispatchers) {
    lateinit var appContext: Context
    actual var lastModified: Instant? = null
```

File: common/src/commonMain/kotlin/com/cuhacking/atlas/common/FeatureApi.kt [hunk 6]
```diff
@@ -3,12 +3,14 @@ package com.cuhacking.atlas.common
...
 import com.cuhacking.atlas.db.Database
 import io.github.dellisd.spatialk.geojson.Feature
-import io.github.dellisd.spatialk.geojson.FeatureCollection.Companion.toFeatureCollection
+import io.github.dellisd.spatialk.geojson.FeatureCollection
 import io.ktor.client.HttpClient
 import io.ktor.client.request.get
 import kotlinx.coroutines.withContext
 
 class FeatureApi(private val database: Database, private val client: HttpClient) {
+    private val dispatchers = CoroutineDispatchers
```
Comment (dellisd) [inline common/src/commonMain/kotlin/com/cuhacking/atlas/common/FeatureApi.kt:12]:
```suggestion
class FeatureApi(private val database: Database, private val client: HttpClient, private val dispatchers: CoroutineDispatchers = CoroutineDispatchers) {
```

File: common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import platform.Foundation.*
+
+actual class DataCache actual constructor(dispatchers: CoroutineDispatchers) {
+    actual var lastModified: Instant? = null
+    private val cacheDirectory = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, true).first()
+    private val file = (cacheDirectory as NSString).stringByAppendingPathComponent("mapdata.json")
+    private val dispatchers = CoroutineDispatchers
```
Comment (dellisd) [inline common/src/iosMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]:
```suggestion
actual class DataCache actual constructor(private val dispatchers: CoroutineDispatchers = CoroutineDispatchers) {
    actual var lastModified: Instant? = null
    private val cacheDirectory = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, true).first()
    private val file = (cacheDirectory as NSString).stringByAppendingPathComponent("mapdata.json")
```

**Review by dellisd — CHANGES_REQUESTED** [review]

**Comment by dellisd** [comment]
Do you plan to add tests for iOS in this PR?
In theory, you should be able to write one common test (although I guess getting that to work with the instrumentation test might be a challenge) and have it run for Android/JVM, JS, and iOS all at once.

File: common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt (hunk 5 above)
Comment (safa-howaid) [reply common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]: I tried to do this earlier, but it kept giving me an error `"Actual function cannot have default argument values, they should be declared in the expected function"`. However, I was able to suppress it and the tests ran successfully. I don't really know why this error came up, should I make the change? (same thing with the rest of the actual classes)

Comment (dellisd) [reply common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]: Ah, in the expect class constructor you have to add the default argument. `expect class DataCache(dispatchers: CoroutineDispatchers = CoroutineDispatchers)`

Comment (safa-howaid) [reply common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]: That was after adding it

Comment (dellisd) [reply common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]: Oh huh 🤔 
Well, we don't need the default argument. We can just specify it when constructing it (even if it seems a little inconvenient).

Comment (safa-howaid) [reply common/src/androidMain/kotlin/com/cuhacking/atlas/common/DataCache.kt:12]: Well...suppressing it worked for some reason, but I don't know if it's going to cause issues later 😆

**Comment by safa-howaid** [comment]
Correct me if I'm wrong, but since this class needs to access the file system on the device, I think the IOS test would need to be run on an emulator to ensure that it's working properly. Most examples of testing on Kotlin MPP involve unit tests, so I think I'll have to look into it some more!

**Review by dellisd — APPROVED** [review]

**Review by ghaskins99 — APPROVED** [review]

**Pull request merged**